specs/Proposals/0040-split-states.md

```diff
@@ -0,0 +1,257 @@
+
+```rust
+pub struct EpochInfoV3 {
+    ... // All fields in EpochInfoV2
+    shards: Vec<ShardId>,
+    parent_shards: HashMap<ShardId, ShardId>,
+}
+```
+
+### ShardVersionManager
+A new struct `ShardVersionManager` will be created to manage sharding assignment when `ProtocolVersion` is changed. `EpochManager` will own `ShardVersionManager` that it can call to create shards for the `EpochInfo` of the next next epoch.
+
+#### ShardVersionManager
+`ShardVersionManager` contains some static fields that store the mapping from `ProtocolVersion` to sharding assignments. Its implementation can be changed easily when a new sharding assignment is added as long as its behavior for the existing protocol versions do not change.
+```rust
+pub const SIMPLE_NIGHTSHADE_SHARD_VERSION: ProtocolVersion;
+pub struct ShardVersionManager {
+    // current shards
+    shards: Vec<ShardId>,
+    // protocol version of the current shards
+    protocol_version: ProtocolVersion,
+}
+```
+#### new
+```rust
+pub fn new(current_protocol_version: ProtocolVersion) -> Self
+```
+creates a new `ShardVersionManager` struct given the current protocol version.
+#### `try_update_shards`
+```rust
+pub fn try_update_shards(protocol_version: ProtocolVersion) -> (Option<(Vec<ShardId>, HashMap<ShardId, ShardId>))
+```
+checks if the given `protocol_version` will trigger a change in sharding. If so, returns the new shards and a mapping from the current shards to parent shards. This will be called in `EpochManager::finalize_epoch` when `next_version` is determined.
+
+### ShardTracker
+Various functions such as `account_id_to_shard_id` in `ShardTracker` will be changed to incorporate the change in `ShardId`. New functions such as `account_id_to_shard_ord` can also be added easily if needed. The `num_shards` field will be removed from `ShardTracker` since it is no longer a static number. The current shard information can be accessed by the following functions. These changes will be propagated to `RuntimeAdapter` since `ShardTracker` cannot be directly accessed through `RuntimeAdapter`.
+
+#### `get_shards`
+```rust
+pub fn get_shards() -> Vec<ShardId>
+```
+returns the `shard_id`s of the shards in the current epoch
+
+#### `get_shards_next_epoch`
+```rust
+pub fn get_shards_next_epoch() -> Vec<ShardId>
+```
+returns the `shard_id`s of the shards in the next epoch
+
+## Build New States
+The following method in `Client` will be added or modified to split a shard's current state into multiple states.
+
+### `split_shards`
+```rust
+pub fn split_shards(me: &Option<AccountId>, sync_hash: CryptoHash, shard_id: ShardId)
+```
+builds states for the new shards that the shard `shard_id` will be split to. After this function is finished, the states for the new shards should be ready in `ShardTries` to be accessed.
```

Comment:
What does this function do exactly? We cannot afford to synchronously compute the new states in this function. Otherwise the client actor will be blocked for a long time

Reply:
I assumed run_catchup is executed in its own thread, so I thought that wouldn't affect other activities of the client actor. I'll take a closer look

Reply:
discussed offline and agreed that we should move the catchup process to its own actor https://github.com/near/nearcore/issues/3201